src/App/VPSScan/NinjaGraph.hs

```diff
@@ -0,0 +1,234 @@
+    NoNinjaDepsEndLineFound   -> "The output of \"ninja -t deps\" did not contain the \"build completed successfully\" line"
+    NinjaDepsParseError   -> "There was an error while parsing the output of \"ninja -t deps\""
+
+data NinjaParseState = Starting | Parsing | Complete | Error
+
+ninjaGraphMain :: NinjaGraphCmdOpts -> IO ()
+ninjaGraphMain NinjaGraphCmdOpts{..} = do
+  dir <- validateDir ninjaCmdBasedir
+  result <- runDiagnostics $ getAndParseNinjaDeps dir ninjaCmdNinjaGraphOpts
+  case result of
+    Left failure -> do
+      print $ renderFailureBundle failure
+      exitFailure
+    Right _ -> pure ()
+
+
+getAndParseNinjaDeps :: (Has Diagnostics sig m, MonadIO m) => Path Abs Dir -> NinjaGraphOpts -> m ()
+getAndParseNinjaDeps dir ninjaGraphOpts = do
+  ninjaDepsContents <- runTrace $ runReadFSIO $ runExecIO $ getNinjaDeps dir ninjaGraphOpts
+  graph <- scanNinjaDeps ninjaGraphOpts ninjaDepsContents
+  _ <- runHTTP $ postDepsGraphResults ninjaGraphOpts graph
+  pure ()
+
+-- If the path to an already generated ninja_deps file was passed in (with the --ninjadeps arg), then
+-- read that file to get the ninja deps. Otherwise, generate it with
+-- NINJA_ARGS="-t deps" make
+getNinjaDeps :: (Has ReadFS sig m, Has Diagnostics sig m, Has Trace sig m, MonadIO m) => Path Abs Dir -> NinjaGraphOpts -> m ByteString
+getNinjaDeps baseDir opts@NinjaGraphOpts{..} =
+  case ninjaGraphNinjaPath of
+    Nothing -> BL.toStrict <$> generateNinjaDeps baseDir opts
+    Just ninjaPath -> readNinjaDepsFile ninjaPath
+
+scanNinjaDeps :: (Has Diagnostics sig m) => NinjaGraphOpts -> ByteString -> m [DepsTarget]
+scanNinjaDeps NinjaGraphOpts{..} ninjaDepsContents =
+  addInputsToNinjaDeps <$> ninjaDeps
+  where
+    ninjaDeps = parseNinjaDeps ninjaDepsContents
+
+depsGraphEndpoint :: Url 'Https -> Url 'Https
+depsGraphEndpoint baseurl = baseurl /: "depsGraph"
+
+-- post the Ninja dependency graph data to the "Dependency graph" endpoint on Scotland Yard
+-- POST /depsGraph
+postDepsGraphResults :: (ToJSON a, MonadIO m, Has Diagnostics sig m) => NinjaGraphOpts -> a -> m ()
+postDepsGraphResults NinjaGraphOpts{..} depsGraph = runHTTP $ do
+  (baseUrl, baseOptions) <- parseUri depsGraphScotlandYardUrl
+  _ <- req POST (depsGraphEndpoint baseUrl) (ReqBodyJson depsGraph) ignoreResponse (baseOptions <> header "Content-Type" "application/json")
+  pure ()
+
+readNinjaDepsFile :: (Has Trace sig m, Has ReadFS sig m, Has Diagnostics sig m, MonadIO m) => FilePath -> m ByteString
+readNinjaDepsFile ninjaPath = do
+  trace $ "reading ninja deps from " ++ ninjaPath
+  path <- liftIO $ parseAbsFile ninjaPath
+  readContentsBS path
+
+generateNinjaDeps :: (Has Trace sig m, Has Diagnostics sig m, MonadIO m) => Path Abs Dir -> NinjaGraphOpts -> m BL.ByteString
+generateNinjaDeps baseDir NinjaGraphOpts{..} = do
+  trace $ "Generating ninja deps with this command: " ++ commandString
```

Comment:
Is this `trace` meant to be left in? This seems like it would be helpful to the user if it was. I'm also not sure if `trace` is the ideal way for us to be logging, I don't have an alternative, I just don't recall using this in the past.

Reply:
Yeah, I was planning on leaving it in.

I've been using `trace` all over the place in this code, and I think it was Connor's suggestion to use it. He's approved PRs that use it at the very least :)